### PR TITLE
fix: Table-wide edge batching in the OpenGraph convert source - BED-9372

### PR DESCRIPTION
**Pull request by ktstrader**
## Summary

The `opengraph` convert source reset its edge-batch accumulator on every source row, so each row emitted its own edge wrapper regardless of `batch_size`. The accumulator now spans the whole graph resource — across rows, DLT `read_jsonl` chunk boundaries, and input files — flushing one final partial batch at end-of-table. A table of `N` edges emits exactly `ceil(N / batch_size)` wrappers (default `batch_size` = 150) instead of one per row.

Framework optimization only: the flattened relationship sequence (order, duplicates, per-edge content) and all nodes are byte-for-byte identical to before. Only edge grouping changes.

Follow-up (same ticket): peak process RSS scaled linearly with table size after batching (~1.2 KB per relationship; 1.26 GB at 1M rows), because DLT's item-count writer buffer now held up to 150× more data per item. The buffer is now coordinated with `batch_size` to keep buffered relationships bounded. Coordinating it exposed a DLT 1.26.0 defect that silently duplicated delivered items; that defect is corrected in-process. Peak RSS is flat post-fix (499 MB at 1M vs 498 MB at 4M rows).

## Motivation

Resolved: [BED-9372](https://specterops.atlassian.net/browse/BED-9372?issueKey=BED-9372&subProduct=jira-software)

## Changes

- `src/openhound/sources/opengraph/source.py`: accumulator moved outside the per-row loop; single final flush; `batch_size < 1` rejected; nodes unchanged and never mixed into edge wrappers.
- `tests/test_opengraph_batching.py`: 19-test regression suite.
- `benchmarks/`: standalone synthetic benchmark harness (not run under pytest) plus a destination memory/part-size review.
- `src/openhound/sources/opengraph/source.py` + `src/openhound/core/convert.py`: `writer_buffer_max_items()` scales `DATA_WRITER__BUFFER_MAX_ITEMS` with `batch_size` (333 @ 150), applied via `setdefault` so user overrides win.
- `src/openhound/core/dlt_jsonl_batching.py` (wired in `openhound/__init__.py`): corrects dlt 1.26.x's `get_batches`, which re-yielded accumulated items per load-file line and duplicated delivery for any non-aligned buffer. Installs upstream's later corrected semantics; no-op on other versions. Note for reviewers: this intentionally modifies a pinned dependency because the alternative is silent duplicate ingestion — stock settings avoid the bug only by coincidence (5,000 divides by 1,000), so any future buffer/batch change would corrupt data with no error raised.
- `benchmarks/_bench_run.py`: records `writer_buffer_max_items` / `peak_rss_per_edge`; RSS guard band (300 MiB + 512 B/edge) flags staging memory that scales with cardinality.
- `benchmarks/DESTINATION_MEMORY_REVIEW.md`: root cause, defect, mitigation, and measured results (RSS flat 1M → 4M; callbacks bounded at ≤150k relationships / ~27 MB).
- `tests/test_writer_buffer.py`, `tests/test_dlt_jsonl_batching.py`: coverage for both.

## Guarantees

- Flattened edges identical to per-row output (same order, duplicates, content; no dedup).
- Wrappers = `ceil(total_edges / batch_size)`, table-wide.
- `batch_size = 1` reproduces per-row wrapping; `batch_size < 1` raises `ValueError`.
- Each resource uses a fresh accumulator; a failed extraction never leaks state into a retry.

Tradeoff: a mid-table failure re-extracts the whole table rather than resuming mid-chunk (documented inline).

Dependency note: the jsonl batching correction is scoped to dlt 1.26.x (the only affected series; upstream fixed it in 1.27.0) and is a no-op elsewhere. Reviewers bumping the dlt pin should confirm the target is 1.27.0+ or still covered by this module.

## Testing

```
.venv\Scripts\python.exe -m pytest tests/test_opengraph_batching.py -v
```

Expect 19 passed — covers cross-row/chunk/file batching, the 1,000-row chunk boundary, `ceil(N/batch_size)` counts, order/duplicate parity vs a `batch_size=1` baseline, edge cases, and per-resource/per-retry isolation.

```
.venv\Scripts\python.exe -m pytest tests/test_writer_buffer.py tests/test_dlt_jsonl_batching.py tests/test_bhe_job_scheduling.py -q
```

Expect 24 passed — buffer coordination bounds and env handling, exact-once jsonl delivery (multi-line files, resume, skipped columns), and BHE ingestion with the coordinated buffer active.

Benchmark (`benchmarks/opengraph_batching_benchmark.py`, one-edge shape): tuned peak RSS 240 MB @ 100k / 499 MB @ 1M / 498 MB @ 4M rows — flat; untuned run reproduces 1.26 GB @ 1M and trips the guard band.

Real-data parity (local, no data committed): replayed against Okta `ApplicationUser` (8,577 edges → 58 wrappers) and GitHub `RepoRoleAssignment` (2,745 edges → 19 wrappers, crosses the chunk boundary, non-Okta). Fixed path, `batch_size=1` baseline, and frozen output all yield identical canonical SHA-256, with model/lookup/extras held constant.

[BED-9372]: https://specterops.atlassian.net/browse/BED-9372?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **New Features**
  - OpenGraph processing now batches relationships across rows, files, chunks, and resources for more efficient large-scale processing.
  - Added validation for batch sizes below 1.
  - Added benchmark tooling to measure batching, memory usage, runtime, and output behavior across graph shapes and dataset sizes.

- **Bug Fixes**
  - Preserved relationship order and duplicates while keeping node and edge output separate.
  - Improved handling of empty inputs, retries, failures, and final partial batches.
  - Prevented invalid configuration values from being silently ignored.

### CODE REVIEW

File: benchmarks/_bench_assets.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        )
+
+    @property
+    def edges(self):
+        return [
+            Edge(
+                kind="BENCH_Relationship",
+                start=EdgePath(match_by="id", value=f"node-n{self.idx}"),
+                end=EdgePath(match_by="id", value=f"end-{self.idx}-0"),
+            )
```
[the remark this answers is not in the document]
Comment (ktstrader) [reply benchmarks/_bench_assets.py:93]: Edge.properties already has field(default_factory=EdgeProperties) with all-default fields (entries_dataclass.py:89). Verified empirically: node_and_edge rows construct and serialize with a valid default properties=EdgeProperties(composed=False, traversable=False). Nothing fails validation.

**Comment by ktstrader** [comment]
@coderabbitai review
